### ISSUE TITLE
copy-paste not working in firefox

### ISSUE THREAD

**Issue by a-x-** [ajaxorg/ace#2794]
hi,

are know about this problem?

**Comment by nightwing** [comment]
Hi. We didn't know about this problem.
Are you talking about copy-paste with keyboard shortcuts or context menu?
Did you test this on ace.c9.io or some other site?
Which OS did you use?

**Comment by a-x-** [comment]
os x maverix
firefox 42

for example, ace on github

and hotkeys and context menu

on screenshot marked text, which will be paste

![2015-12-10 10-07-42 screenshot](https://cloud.githubusercontent.com/assets/6201068/11709143/31cf322a-9f26-11e5-8714-8a9197301464.png)

**Comment by lennartcl** [comment]
I tried this on OSX with version 40.0 and 43.0b9. Copy-paste with keyboard and the context menu both appear to work. The only thing I notice that is slightly off is the "Search Google for " menu item also shown in your screenshot. Other than that if you can't copy-paste that might be caused by a plugin. Could you try again in safe mode? (https://support.mozilla.org/en-US/kb/troubleshoot-firefox-issues-using-safe-mode)

**Comment by a-x-** [comment]
i tried now. with same result...

**Comment by nightwing** [comment]
Do you have a script or a program installed which enhances system clipboard? (e.g adds clipboard history)

**Comment by a-x-** [comment]
nope

my apps:
```
brew cask list
adium			evernote		jing			qlstephen		spectacle		webpquicklook
alfred			firefox			ngrok			quicklook-csv		sublime-text3		webstorm
asepsis			gimp			opera			quicklook-json		suspicious-package	xmind
betterzipql		google-chrome		pomotodo		screenhero		todoist (!)		xquartz
brain-workshop		google-chrome-canary	qlcolorcode		sketch			transmission		yandex
caffeine		iterm2			qlmarkdown		skype			vienna			yandex-beta
chromium		iterm2-beta		qlprettypatch		slack			vlc
```

**Comment by aoaaxy** [comment]
**tl;dr:** Workaround for me(linux+firefox 44.0): `dom.event.clipboardevents.enabled`  must be `true` in Firefox (which is the default)

Hi. I'm experiencing a similar situation: copy/paste not working in linux with firefox and here's a superficial explanation on why and the workaround:
https://github.com/rust-lang/rust-playpen/issues/69#issuecomment-155036257

I'm currently trying to figure out a way to have `dom.event.clipboardevents.enabled` set to the non-default value of `false` in Firefox and still have the copy/pasting work fine (copy/pasting works in every other website thus far, except on play.rust-lang.org and https://rustbyexample.com/hello.html which both seem to be using ACE - as far as I could tell(I don't know javascript/html tho, so I might be wrong))

Any hints on how to fix this are more than welcome. Thank you in advance!

**Comment by aoaaxy** [comment]
I get the [exact same](https://github.com/ajaxorg/ace/issues/2794#issuecomment-163520477) on-right-click 'Search Google for'  when `dom.event.clipboardevents.enabled` is `false`

![pic with false](https://cloud.githubusercontent.com/assets/17016209/12857650/9f739be8-cc4b-11e5-809b-4be4a76d9072.png)

**Comment by nightwing** [comment]
Unfortunately ace can not work without clipboard events, it needs copy event to update the hidden textarea.
It is also possible to update value of textarea when selecting text, but that way editor becomes slow when large text is selected.

**Comment by aoaaxy** [comment]
That is understandable of course.
 Is there some way that I can enable the `update value of textarea when selecting text` for myself ? Or can you tell me where to look in the ACE code if I wanted to do this?
 I would really like to have copy/paste work when clipboard events are disabled.

**tl;dr from below:** Is it possible to modify ACE to never use a hidden text area and instead just use the visible text area for copy/pasting?

If I understand you correctly, there's a hidden plaintext area which ACE maps its *colorful multiline editor that is visible* to  when onCopy is triggered(I don't know javascript, so I'm just guessing here) and from that hidden text area the plain text is being fed to firefox/clipboard. Does that mean that whatever I've selected from the *colorful multiline editor that is visible* cannot be used as text/plaintext to be fed to the clipboard? Or if it can, something else is preventing this from working properly (such as, the fact that copying still pulls text from the hidden text area which has nothing selected due to onCopy never being triggered, while clipboard events are disabled)
 In other words, is the selected text from the visible text area not copy-able? in principle. Or, it is, but other ACE events are preventing it from working(even when clipboard events are disabled) because of the way it's implemented trying to update and use the hidden text area for copying? Maybe then I could disable those events, and thus allow firefox to can copy the visible text by itself?

**Comment by nightwing** [comment]
Ace uses non eidtable nodes to render text. Dom nodes are created only for the visible part of the text, and selection is a div behind the text. This allows Ace to be fast, but browser selection doesn't work so we have to use copy event to fix that. https://github.com/ajaxorg/ace/blob/master/lib/ace/keyboard/textinput.js

> Is there some way that I can enable the update value of textarea when selecting text for myself 

this is possible to do, but you'll have to open browser console and run something like `ace.setOption(...)` to enable it, is that an acceptable solution?

**Comment by aoaaxy** [comment]
> this is possible to do, but you'll have to open browser console and run something like ace.setOption(...) to enable it, is that an acceptable solution?

That sounds acceptable. Eager to try it. Thanks in advance!

**Comment by aoaaxy** [comment]
Do you think it might be a good idea to, if ACE can detect that clipboard events are disabled(or for some reason not working), then automatically fallback to [as you said](https://github.com/ajaxorg/ace/issues/2794#issuecomment-180786353) `It is also possible to update value of textarea when selecting text, but that way editor becomes slow when large text is selected.`

**Comment by nightwing** [comment]
It is a good idea, but AFAIK there is no way to detect that.

**Comment by aoaaxy** [comment]
Or here's a random idea, if it's possible for ACE to detect when the right click mouse button is pressed (or when that menu appears) is pressed, then execute `editor.getCopyText();` and have ACE copy that into clipboard?(or maybe in order for ACE to modify clipboard, it does still need `dom.event.clipboardevents.enabled` to be `true`?) Or maybe a new menu item can be added ?
 Ok, I don't know :)

**Comment by aoaaxy** [comment]
> this is possible to do, but you'll have to open browser console and run something like ace.setOption(...) to enable it, is that an acceptable solution?

Can you please tell me which setOption to use there? I'm assuming it's already implemented, otherwise, you don't have to implement a new one just for this, obviously.

Otherwise, I could just use `editor.getCopyText();` for now, it's still better than nothing :D (since I do want to keep clipboard events off for to avoid the case when websites can capture my copy/pasting of the password from the password manager to some website's password text area)

**Comment by nightwing** [comment]
I meant we can implement that, but it won't be convenient to use, because we do not have a way to detect missing support for clipboard events.

Webpages can't modify or read clipboard outside of clipboard events.

> (since I do want to keep clipboard events off for to avoid the case when websites can capture my copy/pasting of the password from the password manager to some website's password text area)

I don't understand. Only the website in which you paste can capture contents of your clipboard.
But when you paste into password text area, you already give that password to the website, and website can detect changes to textarea using input event too.

**Comment by aoaaxy** [comment]
> Only the website in which you paste can capture contents of your clipboard.
But when you paste into password text area, you already give that password to the website, and website can detect changes to textarea using input event too.

My bad! I misunderstood how clipboard events work - I thought that whenever I do a Copy (outside of firefox even), then all websites that have event listeners for onCopy would get notified and could capture the contents of the (system)clipboard. 
 Furthermore I thought websites could also read the clipboard by just having timers and repeatedly checking for clipboard contents.

> Webpages can't modify or read clipboard outside of clipboard events.

Got it! (I'm assuming `clipboard events` are local to each webpage, so a Copy from another webpage won't trigger onCopy in a different webpage and thus also a Copy in another program(outside of firefox) won't trigger any onCopy events in any firefox webpages)

Thank you for this!

**Comment by aoaaxy** [comment]
I want to add a correction to [this comment I made](https://github.com/ajaxorg/ace/issues/2794#issuecomment-180535532):
Even with `dom.event.clipboardevents.enabled` `true` the `Search Google for` is still the same. (even if the clipboard copy/pasting now works). It seems to me that the value for `Search Google for` is taken from the selection (selected text).

**Comment by aoaaxy** [comment]
I would like to add that in `about:config` the key [`clipboard.autocopy`](http://kb.mozillazine.org/Clipboard.autocopy) is set to `true` by default but firefox won't fire a copy event for when it does copy data into the second(?) clipboard(which happens when user is done selecting the text - this, outside of ACE) from which I can only paste via middle click mouse button (in Linux).

I tested this by using https://github.com/ajaxorg/ace/blob/55313882c9e6f686489f5e1c1acf67fd3dfb7c7f/experiments/cut_copy.html and adding an `e` param + a call to `log(e)` in the [fillSelection function](https://github.com/ajaxorg/ace/blob/55313882c9e6f686489f5e1c1acf67fd3dfb7c7f/experiments/cut_copy.html#L81)

Meanwhile I'm trying to figure out how to make ACE keep some kind of (hidden text area) text selection(even when clipboard events are enabled! instead of just that [one char marker](https://github.com/ajaxorg/ace/issues/2794#issuecomment-180535532)) so that maybe firefox can pick that up and use that (selected)text for when displaying the `Search Google for` menu item. 
 I'm thinking that maybe if I can do that, then I can get Firefox to have that text autocopied via the `clipboard.autocopy;true` (when clipboard events are disabled) and thus I can just middle click(to paste) after selecting that text in ACE editor. Wouldn't that be great! :D

**Comment by un1versal** [comment]
I have this issue (for over a year now) and have reported it to GitHub (recently) who as usual dont have any issues whatsoever and everything works for them normally. :-1: 

it was Rachel one of their support personnel which indicated this ticket maybe related but they strongly suggest that thiss is a local issue, despite default installs of OS's and Browsers and settings being untouched.

Both Windows and Linux are affected and so are other Browsers,

I have asked GitHub what technologies they use for this purpose and they said ACE Editor, but alas Im none the wiser, I was wondering what technologies ACE uses to facilitate this but from experience its probably not a good idea to ask such questions.

So bug exists and it goes unfixed and live goes on.

**Comment by nightwing** [comment]
@un1versal this bug is not about different browsers with settings being untouched, this is about Firefox with `dom.event.clipboardevents.enabled` turned off. Could you please tell on which browsers do you see this issue, and what is the value of `dom.event.clipboardevents.enabled` on firefox?

**Comment by un1versal** [comment]
Well like I said I was pointed to this report by GitHub personnel who insist this is a local issue.

The issue I face is also not just about the search for (something) being mangled, it about not being able to copy and paste anything via browser which I reported to GitHub and who pointed me here perhaps as a mean to investigate locally.. (though I also suffer from this issue)

My Firefox (46.x) is set  per default which by definition is settings being untouched.

![capture](https://cloud.githubusercontent.com/assets/3521959/15773465/e48ca974-296e-11e6-9a32-d8b2268796de.PNG)

Internet explorer and Chrome across Linux/Windows all present similar results.

**Comment by nightwing** [comment]
- Do you see the error only in ace or in other textareas too. 
- Do you see it on https://ace.c9.io/#nav=about
- Do you paste with context menu or ctrl-v keyboard shortcut?

**Comment by un1versal** [comment]
* Apparently only in Ace
* Yes I see it there also.
* Doesnt seem to make a difference which method of copying or pasting is used

I can copy and paste say any raw text from GitHub or from any website or anywhere else from source to any destination without any issues.

Ill blame java for now here, I have set the highest security (if there is such a thing with java) in Java control panel > security tab,

**Comment by nightwing** [comment]
@un1versal  I couldn't reproduce this issue either, despite of spending long time on it.

You are saying that in the first group of screenshots copying worked in the `Example on left`, but didn't work in `Example on right`

Example on left shows text being copied from Ace, and Example on right shows text being copied from outside of Ace. Code from Ace can not affect text outside of editable area, so there must be some other issue here.

>  for content in an editor I want to copy is to view raw contents in browser and copy that and paste that anywhere I like

when copying doesn't work on github
do you see this:
![image](https://cloud.githubusercontent.com/assets/341801/15965739/cc9e29ba-2f30-11e6-87ce-e05b26def7e5.png)
or this:
![image](https://cloud.githubusercontent.com/assets/341801/15965758/e2eb8f82-2f30-11e6-9620-58425a68051e.png)
?

**Comment by un1versal** [comment]
> when copying doesn't work on github
>do you see this:

I see this: definitely

![this](https://cloud.githubusercontent.com/assets/341801/15965758/e2eb8f82-2f30-11e6-9620-58425a68051e.png)

**Comment by nightwing** [comment]
And about the first question, is there a typo or did copying not work for text outside of editor?

**Comment by un1versal** [comment]
it works fine outside the editor

yes there was a typo the result 3) was inverted I corrected it now (and confirmed by retest)... sorry.

**Comment by bopoh13** [comment]
FF50 [[cfg](https://github.com/bopoh13/docs/blob/master/vendor/firefox/user.js)], FF51. Do not works the Paste-hotkey combination. Next...

![clipboardevents.enabled](https://cloud.githubusercontent.com/assets/3521959/15773465/e48ca974-296e-11e6-9a32-d8b2268796de.PNG)
After [changing the setting](https://github.com/ajaxorg/ace/issues/2794#issuecomment-180527048) does not work **Shift+Insert** only on GitHub to change the text or create a new file independently. In comments to issues and [https://ace.c9.io/](https://github.com/ajaxorg/ace/issues/2794#issuecomment-223529133) works.

**Comment by bopoh13** [comment]
**The issue can be closed.** The problem is no actual.

**Comment by IceSentry** [comment]
This problem is very much still "actual". I can reproduce every scenario that un1versal showed in his comment. I can't copy anything from anything that uses the ace editor but everything else is fine.

**Comment by Silex** [comment]
Same problem here, cannot copy text using the browser's right click menu. It's pretty annoying to be forced to use CTRL-C or CTRL-V

**Comment by IceSentry** [comment]
Did you mean CTRL-X CTRL-V to copy, or do you actually prefer to right click to copy? For me the right click and CTRL-C doesn't work

**Comment by Silex** [comment]
I mean that CTRL-C/CTRL-X/CTRL-V works. The right click menu does not show any of "Copy", "Cut" or "Paste", as if nothing was selected.

We are using codiad.com to work with students learning IT, and right click makes it simpler to explain than remembering the CTRL shortcuts.

**Comment by IceSentry** [comment]
You might want to open a new issue because this is not the same thing at all. This is about copy paste not working at all. For right click menu or shortcuts, this might be caused by the same thing, but consider yourself lucky that you can at least use the ctrl shortcuts.

Also, I know this is not the place to argue about that, but who is learning IT in 2017 and doesn't know what ctrl-c/ctrl-v does. It's probably the most well known shortcut that is used  pretty much everywhere. I'm not here to judge, but I find it hard to believe that anyone struggle with the concept of using that shortcut.

**Comment by dhardy** [comment]
Yeah, is this about copy/paste not working at all, needing "clipboard events" enabled or the context menu not working? Maybe it would be worth opening a new issue.

For me paste works fine (from another source) but copy doesn't work at all (context or ctrl+C), and "clipboard events" are enabled.

**Comment by nightwing** [comment]
Could you try going to https://rawgit.com/nightwing/ace/t1/experiments/cut_copy.html and trying to follow instruction on that page.

I am suspecting either interference from browser extension, or from secondary clipboard, but it may be something entirely different, and the log output of that page should help us to figure what is going on.

**Comment by dhardy** [comment]
```
7:paste t3 get:=copied=5
6:...defaultPrevented
5:copy t2 set:
4:copy t2 get:
3:paste t3 get:=copied=1
2:...defaultPrevented
1:copy t1 set:
0:copy t1 get:
```

**Comment by nightwing** [comment]
@dhardy thanks! this is very helpful, could you please try going to https://rawgit.com/nightwing/ace/t2/experiments/cut_copy.html and copy text from the ace editor instance into the textarea 3 .

Also in https://github.com/integer32llc/rust-playground/issues/227, you say  `my clipboard is filled with �`. Is it always filled with � after selecting text in ace, or only after trying to copy?

**Comment by dhardy** [comment]
Result of copying from `play.rust-lang.org`:

```
0:input "3\u0001"
```

The clipboard is set when copying (ctrl+C or context menu). Curiously the mouse buffer is not affected by selecting text in the ACE editor (it still contains content from the previous selection).

Note: my other computer with similar OS & browser can copy text from the ACE editor correctly. If you like I can try narrowing down the differences.

**Comment by nightwing** [comment]
Could you please try to copy from the ace instance on the https://rawgit.com/nightwing/ace/t2/experiments/cut_copy.html page.
 
![image](https://user-images.githubusercontent.com/341801/32516503-03bab576-c41d-11e7-9b13-abafc3c9b137.png)

In the version of ace on that page, I have added logging and made a change that possibly can fix the clipboard issue.

> If you like I can try narrowing down the differences.

That would help very helpful, but may be hard to do.

**Comment by dhardy** [comment]
Ok (there's an extra copy then ctrl+z, which are the last two lines).
```
13:input "312x"
12:input "312xx"
11:input "312x"
10:input "312"
9:input "31"
```

**Comment by nightwing** [comment]
That is rather weird, the clipboard events fire for the first textarea, but do not fire for the textarea inside ace.
Do you have any browser extensions installed?

In https://rawgit.com/nightwing/ace/t3/experiments/cut_copy.html i have tried to eliminate all the styling  differences between  textarea, could you try copy pasting with keyboard from ace and from the first textarea on that page, maybe this time it will work.

**Comment by dhardy** [comment]
```
2:input "312x"
1:input "312"
0:input "31"
```

Contrary to what I said earlier, I see that `dom.event.clipboardevents.enabled` was disabled. With that enabled, copy+paste works and I get:
```
24:input "3=copied=1=copied=7=copied=20"
23:setDataError NoModificationAllowedError: Modifications are not allowed for this document
22:paste t3 get:=copied=20
21:...defaultPrevented
20:copy set:ace1
19:copy get:ace1
18:took clipboard path
17:returning true
16:ace1
15:mime is text/plain
14:handleClipboardData: ace1 `ie`
13:data ace1
12:doCopy
11:input "3=copied=1=copied=7"
10:setDataError NoModificationAllowedError: Modifications are not allowed for this document
9:paste t3 get:=copied=7
8:...defaultPrevented
7:copy t2 set:
6:copy t2 get:
5:input "3=copied=1"
4:setDataError NoModificationAllowedError: Modifications are not allowed for this document
3:paste t3 get:=copied=1
2:...defaultPrevented
1:copy t1 set:
0:copy t1 get:
```

**Comment by nightwing** [comment]
Did you disable intentionally or was it enabled by some extension?
Also do you want to you want to keep them disabled, or are you happy with the solution?

Asking because we may be able to add some warning/fallback for this case, but would like to know how useful that is going to be.

**Comment by dhardy** [comment]
Oh; I don't remember. It's possible I did it myself but can't remember why I would have. I'm happy enough now thanks.

As you can see from this issue there are a *few* people who've hit this (+5 on [this comment](https://github.com/ajaxorg/ace/issues/2794#issuecomment-180527048)), but not very many, so you'll have to decide that.

**Comment by hene193** [comment]
This issue is now present on Chrome on OS X. Should I open a new issue or just add to this thread?

**Comment by nightwing** [comment]
@hene193 a new issue would be better.

**Comment by EnricoScantamburlo** [comment]
Hello,
my company application will embed the ace-editor. I was asked to solve this and I tried to debug the problem.
I cannot copy in Firefox, the problem is reproducible here https://ace.c9.io/build/kitchen-sink.html and if you enable the Text Input Debugger you will see that only a part of the text gets copied. 
The results of my debugging was that the listeners of the textarea ( the doCopy() method) are never called. The Firefox does not stop if I put a breakpoint on the onCopy event. It does not work even if add a listeners of mine or add the onClick attribute.

I think it is the same problem as https://github.com/ajaxorg/ace/issues/3976 no events are fired 
It could be because the element if not attached to the DOM? 
https://stackoverflow.com/questions/18303695/dynamically-created-ahref-click-event-not-working-in-firefox-or-safari

Any suggestions?

**Comment by EnricoScantamburlo** [comment]
FYI
My company had **dom.event.clipboardevents.enabled** set to **false** by default. Enabling it just fixed the problem. I was wondering if using the new ClipboardApi the situation could improve.

**Comment by nightwing** [comment]
Sadly firefox doesn't support the `readText` method of the new api too https://developer.mozilla.org/en-US/docs/Web/API/Clipboard/readText#browser_compatibility
